bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/AbstractLinkRegistry.java

```diff
@@ -25,16 +31,89 @@
  *
  * @author Dennis Nobel - Initial contribution
  *
- * @param <L>
- *            Concrete type of the abstract link
+ * @param <L> Concrete type of the abstract link
  */
 public abstract class AbstractLinkRegistry<L extends AbstractLink, P extends Provider<L>>
         extends AbstractRegistry<L, String, P> {
 
+    private final ReentrantReadWriteLock toLinkLock = new ReentrantReadWriteLock();
+    private final Map<String, Set<L>> itemNameToLink = new HashMap<>();
+    private final Map<UID, Set<L>> linkedUidToLink = new HashMap<>();
+
     protected AbstractLinkRegistry(final Class<P> providerClazz) {
         super(providerClazz);
     }
 
+    @Override
+    protected void onAddElement(final L element) {
+        toLinkAdded(element);
+    }
+
+    @Override
+    protected void onRemoveElement(final L element) {
+        toLinkRemoved(element);
+    }
+
+    @Override
+    protected void onUpdateElement(final L oldElement, final L element) {
+        toLinkRemoved(oldElement);
+        toLinkAdded(element);
+    }
+
+    private void toLinkAdded(final L element) {
+        final String itemName = element.getItemName();
+        final UID linkedUid = element.getLinkedUID();
+
+        toLinkLock.writeLock().lock();
+        try {
+            Set<L> set;
+
+            set = itemNameToLink.get(itemName);
+            if (set == null) {
```

Comment:
This part can be simplified with `computeIfAbsent`: https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-

Reply:
Sure, but does it matter? :wink: 

We create a new lambda expression (the argument for `computeIfAbsent`) and the default implementation of `computeIfAbsent` adds other additional method call (at least the function of the lambda expression).

It is the `add` method that is called for every element in of a provider if it is added to the registry.
So wouldn't it make sense to try to keep it as performant as possible?

Reply:
WRT the test code above:
Adding a provider that contains 10000 entries to a registry already took 1 minute on my working machine because `added` (and `on...`) is called for every element.

Reply:
I did some performance tests with the `computeIfAbsent` using your test and could no see a difference. Here is the code I used. So do what you think is best:
```
            itemNameToLink.computeIfAbsent(itemName, k -> new HashSet<>(1)).add(element);
            linkedUidToLink.computeIfAbsent(linkedUid, k -> new HashSet<>(3)).add(element);
```
P.s I did some premature optimization here (initialized HashSet with small numbers (default size is 16), assuming the size of these set will be in general very small I think)

Reply:
Years ago I developed for embedded devices in C / C++ for years.
Every function call, ever check against null, ... influences the runtime.
Functions are inlined just to improve the performance...

I wonder all the time why we don't care on Java anymore.
The default implementation will call the provided lambda function, it will check if the return value differs from null (we already know that `new ...` will not return null)... If I want to write more performant code I don't care about additional lines of code but if I can reduce checks and function calls.
Perhaps there is some magic in Java I don't know but IMHO every check will consume some time. On embedded devices such checks will consume more time then on development machines.
If you think it does not matter and we don't notice it on small embedded devices, I can change the code.

Reply:
The change could look like:

```diff
diff --git a/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/AbstractLinkRegistry.java b/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/AbstractLinkRegistry.java
index b6d413d78f..63d71cdfbf 100644
--- a/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/AbstractLinkRegistry.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/AbstractLinkRegistry.java
@@ -18,6 +18,7 @@ import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 
 import org.eclipse.smarthome.core.common.registry.AbstractRegistry;
@@ -35,6 +36,7 @@ import org.eclipse.smarthome.core.thing.UID;
 public abstract class AbstractLinkRegistry<L extends AbstractLink, P extends Provider<L>>
         extends AbstractRegistry<L, String, P> {
 
+    private final Function<Object, Set<L>> toLinkValueSetCreator = unused -> new HashSet<>();
     private final ReentrantReadWriteLock toLinkLock = new ReentrantReadWriteLock();
     private final Map<String, Set<L>> itemNameToLink = new HashMap<>();
     private final Map<UID, Set<L>> linkedUidToLink = new HashMap<>();
@@ -70,18 +72,10 @@ public abstract class AbstractLinkRegistry<L extends AbstractLink, P extends Pro
         try {
             Set<L> set;
 
-            set = itemNameToLink.get(itemName);
-            if (set == null) {
-                set = new HashSet<>();
-                itemNameToLink.put(itemName, set);
-            }
+            set = itemNameToLink.computeIfAbsent(itemName, toLinkValueSetCreator);
             set.add(element);
 
-            set = linkedUidToLink.get(linkedUid);
-            if (set == null) {
-                set = new HashSet<>();
-                linkedUidToLink.put(linkedUid, set);
-            }
+            set = linkedUidToLink.computeIfAbsent(linkedUid, toLinkValueSetCreator);
             set.add(element);
         } finally {
             toLinkLock.writeLock().unlock();
```

But it should result in the following overhead (Java 8) per map:
* additional function call: `Objects.requireNonNull(mappingFunction);`
* additional check in that function: `obj == null`
* one additional function call: `mappingFunction.apply(key)`
* one additional check: `(newValue = ...) != null`